### PR TITLE
[wwb] Similarity for reranking models become based on relevance scores

### PR DESCRIPTION
**Pull request by sbalandi**
## Description
- set def_top_k as very big(todo: add cmd option)
- remove check of ranking order

similarity with fp16:
similarity(int8_asym) 0.935997
similarity(int8_sym)  0.943106
similarity(int4_data_aware) 0.773044
similarity(int4_data_free) 0.497591
similarity(int4_data_free_all_layers) 0.493689

Ticket: CVS-174854

## Checklist:
- [ ] Tests have been updated or added to cover the new code 
- [ ] This patch fully addresses the ticket. 
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: tools/who_what_benchmark/whowhatbench/reranking_evaluator.py [hunk 1]
```diff
@@ -164,11 +164,10 @@ def default_gen_answer(model, tokenizer, query, passages):
...
                     scores = outputs.flatten()
                 scores = scipy.special.expit(scores)
-            sorted_scores = []
+            ordered_scores = []
             for index, (score, _) in enumerate(zip(scores, passages)):
-                sorted_scores.append(np.array([index, score.numpy()]))
-            sorted_scores.sort(key=lambda x: x[1], reverse=True)
-            return np.array(sorted_scores[:DEFAULT_TOP_K])
+                ordered_scores.append(np.array([index, score.numpy()]))
+            return np.array(ordered_scores[:DEFAULT_TOP_K])
```
Comment (as-suvorov) [inline tools/who_what_benchmark/whowhatbench/reranking_evaluator.py:173]: Do we need `DEFAULT_TOP_K` at all? We need all documents to be returned, maybe we don't need this slicing at all?
For genai reranker pipeline we have `top_n=3` by default, do we pass `top_n` for genai pipeline additionally?

File: tools/who_what_benchmark/whowhatbench/whowhat_metrics.py [hunk 2]
```diff
@@ -216,18 +215,16 @@ def evaluate(self, data_gold, data_prediction):
 
             per_query_text = []
             for i, score in enumerate(gold_data):
```
Comment (as-suvorov) [inline tools/who_what_benchmark/whowhatbench/whowhat_metrics.py]: `i` seems to be not used
```suggestion
            for document_idx, gold_score in gold_data:
```

File: tools/who_what_benchmark/whowhatbench/whowhat_metrics.py [hunk 3]
```diff
@@ -216,18 +215,16 @@ def evaluate(self, data_gold, data_prediction):
...
-                # documets on the same position of top_n is different
-                if i >= len(prediction_data) or int(score[0]) != int(prediction_data[i][0]):
-                    per_query_text.append(math.inf)
-                else:
-                    per_query_text.append(abs(score[1] - prediction_data[i][1]))
+                document_idx = score[0]
+                pred_i = 0
+                for pred_i, pred_score in enumerate(prediction_data):
+                    if int(pred_score[0]) == int(document_idx):
```
Comment (as-suvorov) [inline tools/who_what_benchmark/whowhatbench/whowhat_metrics.py]: Valid comment from copilot. If document id is not found then something went wrong, lets fail in that case

File: tools/who_what_benchmark/whowhatbench/whowhat_metrics.py [hunk 4]
```diff
@@ -216,18 +215,16 @@ def evaluate(self, data_gold, data_prediction):
...
+                        per_query_text.append(abs(score[1] - pred_score[1]))
+                        break
+                prediction_data = np.delete(prediction_data, pred_i, 0)
```
[the remark this answers is not in the document]
Comment (as-suvorov) [reply tools/who_what_benchmark/whowhatbench/whowhat_metrics.py]: Please remove deletion or use dictionary if you want to optimize perf

**Comment by as-suvorov** [comment]
> ## Description
> * set def_top_k as very big(todo: add cmd option)
> * remove check of ranking order
> 
> similarity with fp16: similarity(int8_asym) 0.935997 similarity(int8_sym) 0.943106 similarity(int4_data_aware) 0.773044 similarity(int4_data_free) 0.497591 similarity(int4_data_free_all_layers) 0.493689
> 
> Ticket: [CVS-174854](https://jira.devtools.intel.com/browse/CVS-174854)
> 
> ## Checklist:
> * [ ]  Tests have been updated or added to cover the new code
> * [ ]  This patch fully addresses the ticket.
> * [ ]  I have made corresponding changes to the documentation

Did you check logits similarity?
cc: @andreyanufr

File: tools/who_what_benchmark/whowhatbench/reranking_evaluator.py (hunk 1 above)
Comment (sbalandi) [reply tools/who_what_benchmark/whowhatbench/reranking_evaluator.py:173]: yes, top_k is mainly needed for genAI, is it ok to compare only 3 in genAI case ?

Comment (sbalandi) [reply tools/who_what_benchmark/whowhatbench/reranking_evaluator.py:173]: In general, there are about 10 documents in the default example

Comment (as-suvorov) [reply tools/who_what_benchmark/whowhatbench/reranking_evaluator.py:173]: Let's compare all documents which were passed as arguments - always set top_k == number of documents

File: tools/who_what_benchmark/whowhatbench/whowhat_metrics.py (hunk 2 above)
Comment (sbalandi) [reply tools/who_what_benchmark/whowhatbench/whowhat_metrics.py]: updated

File: tools/who_what_benchmark/whowhatbench/whowhat_metrics.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sbalandi) [reply tools/who_what_benchmark/whowhatbench/whowhat_metrics.py]: removed

File: tools/who_what_benchmark/whowhatbench/whowhat_metrics.py [hunk 5]
```diff
@@ -214,20 +213,17 @@ def evaluate(self, data_gold, data_prediction):
...
-                # documets on the same position of top_n is different
-                if i >= len(prediction_data) or int(score[0]) != int(prediction_data[i][0]):
-                    per_query_text.append(math.inf)
-                else:
-                    per_query_text.append(abs(score[1] - prediction_data[i][1]))
-            metric_per_query.append(per_query_text)
+            for document_idx, gold_score in gold_data:
+                scores_diff = 1
+                if document_idx in prediction_scores:
+                    scores_diff = abs(gold_score - prediction_scores[document_idx])
```
[the remark this answers is not in the document]
Comment (as-suvorov) [reply tools/who_what_benchmark/whowhatbench/whowhat_metrics.py]: Good point. @sbalandi could you please add comments for both `scores_diff=1` and default_top_k

[the remark this answers is not in the document]
Comment (sbalandi) [reply tools/who_what_benchmark/whowhatbench/whowhat_metrics.py]: added comments and constant
